services/sources/dspace_service.py

```diff
@@ -0,0 +1,131 @@
+                    continue
+
+                try:
+                    parsed_record = self.parse_record(record)
+                    mapped_records.append(parsed_record)
+                except DSpaceError as e:
+                    logger.error(f'Error parsing DSpace record {record}')
+
+                records_processed += 1
+
+                if limit is not None and records_processed >= limit:
+                    return mapped_records
+
+        return mapped_records
+
+    def parse_record(self, record):
+        try:
+            record = self.source_mapping(record, self.OAI_NAMESPACES, self.constants)
+            record.applyMapping()
+            return record
+        except MappingError as e:
+            raise DSpaceError(e.message)
+
+    def get_single_record(self, record_id, source_identifier):
+        url = f'{self.base_url}verb=GetRecord&metadataPrefix=oai_dc&identifier={source_identifier}:{record_id}'
+
+        response = requests.get(url, timeout=30)
+
+        if response.status_code == 200:
+            content = BytesIO(response.content)
+            oaidc_XML = etree.parse(content)
+            oaidc_record = oaidc_XML.xpath('//oai_dc:dc', namespaces=self.OAI_NAMESPACES)[0]
+
+            try:
+                parsed_record = self.parse_record(oaidc_record)
+                return parsed_record
+            except DSpaceError as e:
+                logger.error(f'Error parsing DSpace record {oaidc_record}')
+
+    def get_resumption_token(self, oai_file):
+        try:
+            oai_XML = etree.parse(oai_file)
+            return oai_XML.find('.//resumptionToken', namespaces=self.ROOT_NAMESPACE).text
+        except AttributeError:
+            return None
+
+    def download_records(self, full_import, start_timestamp, resumption_token=None):
+        headers = {
+            # Pass a user-agent header to prevent 403 unauthorized responses from DSpace
+            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3"
+        }
+
+        url_params = 'verb=ListRecords'
+        if resumption_token:
+            url_params = f'{url_params}&resumptionToken={resumption_token}'
+        elif full_import is False:
+            if not start_timestamp:
+                start_timestamp = (datetime.now(timezone.utc).replace(
+                    tzinfo=None) - timedelta(hours=24)).strftime('%Y-%m-%d')
+            url_params = f'{url_params}&metadataPrefix=oai_dc&from={start_timestamp}'
+        else:
+            url_params = f'{url_params}&metadataPrefix=oai_dc'
+
+        url = f'{self.base_url}{url_params}'
+
+        response = requests.get(url, stream=True, timeout=30, headers=headers)
+
+        if response.status_code == 200:
+            content = bytes()
+
+            for chunk in response.iter_content(1024 * 100):
+                content += chunk
+
+            return BytesIO(content)
+
+        raise DSpaceError(
```

Comment:
I think we can just raise an Exception here unless downstream we need to handle it in a specific way.